roles/lib_yaml_editor/build/src/yedit.py

```diff
@@ -47,20 +75,52 @@ def add_entry(data, keys, item):
             keys = a.b
             item = c
         '''
+        if not isinstance(data, dict) and not isinstance(data, list):
+            return data
+
         if "." in keys:
             key, rest = keys.split(".", 1)
-            if key not in data:
-                data[key] = {}
 
-            if not isinstance(data, dict):
-                raise YeditException('Invalid add_entry called on a [%s] of type [%s].' % (data, type(data)))
-            else:
-                Yedit.add_entry(data[key], rest, item)
+            arr = Yedit.arr_index_re.findall(key)
+
+            if isinstance(data, list):
+                if arr:
+                    Yedit.add_entry(data[int(arr[0])], rest, item)
+
+                elif key not in data:
+                    data[key] = {}
+
+                else:
+                    Yedit.add_entry(data[key], rest, item)
+
+            elif isinstance(data, dict):
+                if arr:
+                    key = keys.split('[')[0]
+                    Yedit.add_entry(data[key][int(arr[0])], rest, item)
+
+
+                else:
+                    if key not in data:
+                        data[key] = {}
+
+                    Yedit.add_entry(data[key], rest, item)
 
         else:
-            data[keys] = item
+            arr = Yedit.arr_index_re.findall(keys)
+            if isinstance(data, list):
+                if arr:
+                    data[int(arr[0])] = item
+
+            elif isinstance(data, dict):
+                if arr:
+                    keys = keys.split('[')[0]
+                    data[keys][int(arr[0])] = item
+
+                else:
+                    data[keys] = item
 
 
+    # pylint: disable=too-many-branches,too-many-return-statements
     @staticmethod
     def get_entry(data, keys):
```

Comment:
get_entry, delete_entry and add_entry container a lot of the same functionality to find the specific part of the dictionary to act on. That code needs to be shared. Here it's repeated over and over.